File: php/src/Resources/Conversions.php
<?php

declare(strict_types=1);

namespace Rerout\Resources;

use Rerout\Models\RecordConversionInput;
use Rerout\Models\RecordedConversion;
use Rerout\Rerout;

/**
 * Conversion tracking namespace. Reached via `Rerout::conversions()`.
 *
 * Records conversion events against a click so attribution and value can be
 * reported back in analytics.
 */
final class Conversions
{
    public function __construct(private readonly Rerout $client)
    {
    }

    /**
     * Record a conversion event for a click. Idempotent — recording the same
     * click + event twice returns `duplicate = true` and keeps the first record.
     */
    public function record(RecordConversionInput $input): RecordedConversion
    {
        /** @var array<string, mixed> $response */
        $response = $this->client->request(
            method: 'POST',
            path: '/v1/conversions',
            body: $input->toArray(),
        );

        return RecordedConversion::fromArray($response);
    }
}

File: dotnet/src/Rerout/Resources/Conversions.cs
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using Rerout.Internal;
using Rerout.Models;

namespace Rerout.Resources;

/// <summary>
/// Conversion tracking: record conversions against prior clicks. Reached through
/// <see cref="ReroutClient.Conversions"/>.
/// </summary>
public sealed class Conversions
{
    private readonly ReroutHttpHandler _http;

    internal Conversions(ReroutHttpHandler http) => _http = http;

    /// <summary>
    /// Record a conversion attributed to a prior click via its <c>ClickId</c>
    /// (<c>rrid</c>). Idempotent per <c>(ClickId, EventName)</c>: a repeat call
    /// returns <see cref="RecordedConversion.Duplicate"/> set to <c>true</c>
    /// without double-counting.
    /// </summary>
    /// <param name="input">The conversion to record. <c>ClickId</c> and <c>EventName</c> are required.</param>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    /// <returns>Whether the conversion was recorded and whether it was a duplicate.</returns>
    /// <exception cref="ReroutException">On any API or transport failure.</exception>
    public Task<RecordedConversion> RecordAsync(
        RecordConversionInput input,
        CancellationToken cancellationToken = default)
    {
        return _http.SendAsync<RecordedConversion>(
            new ReroutRequest
            {
                Method = HttpMethod.Post,
                Path = "/v1/conversions",
                Body = input,
            },
            cancellationToken);
    }
}
